test/E2ETest/Microsoft.Test.E2E.AspNet.OData/ModelBuilder/RecursiveComplexTypesTests.cs

```diff
@@ -0,0 +1,489 @@
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+using Xunit;
+
+namespace Microsoft.Test.E2E.AspNet.OData.ModelBuilder
+{
+    public class RecursiveComplexTypesTests_UserEntity
+    {
+        public int ID { get; set; }
+
+        public RecursiveComplexTypesTests_Customer Customer { get; set; }
+
+        public RecursiveComplexTypesTests_Address Address { get; set; }
+
+        public RecursiveComplexTypesTests_Directory HomeDirectory { get; set; }
+
+        public List<RecursiveComplexTypesTests_Field> CustomFields { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests_JustCustomer
+    {
+        public int ID { get; set; }
+
+        public RecursiveComplexTypesTests_Customer Customer { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests_JustAddress
+    {
+        public int ID { get; set; }
+
+        public RecursiveComplexTypesTests_Address Address { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests_JustHomeDirectory
+    {
+        public int ID { get; set; }
+
+        public RecursiveComplexTypesTests_Directory HomeDirectory { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests_JustCustomFields
+    {
+        public int ID { get; set; }
+
+        public List<RecursiveComplexTypesTests_Field> CustomFields { get; set; }
+    }
+
+    // Scenario 1: Direct reference (complex type points to itself)
+    public class RecursiveComplexTypesTests_Address
+    {
+        public string Street { get; set; }
+
+        public string City { get; set; }
+
+        public string Country { get; set; }
+
+        public RecursiveComplexTypesTests_Address PreviousAddress { get; set; }
+    }
+
+    // Scenario 2: Collection (complex type points to itself via a collection)
+    public class RecursiveComplexTypesTests_Field
+    {
+        public string Name { get; set; }
+
+        public string DataType { get; set; }
+
+        public List<RecursiveComplexTypesTests_Field> SubFields { get; set; }
+    }
+
+    // Scenario 3: Composition (complex type points to itself via indirect recursion)
+    public class RecursiveComplexTypesTests_Customer
+    {
+        public string Name { get; set; }
+
+        public List<RecursiveComplexTypesTests_Account> Accounts { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests_Account
+    {
+        public int Number { get; set; }
+
+        public RecursiveComplexTypesTests_Customer Owner { get; set; }
+    }
+
+    // Scenario 4: Inheritance (complex type has sub-type that points back to the base type via a collection)
+    public class RecursiveComplexTypesTests_File
+    {
+        public string Name { get; set; }
+
+        public int Size { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests_Directory : RecursiveComplexTypesTests_File
+    {
+        public List<RecursiveComplexTypesTests_File> Files { get; set; }
+    }
+
+    public class RecursiveComplexTypesTests : WebHostTestBase
```

Comment:
>RecursiveComplexTypesTests [](start = 17, length = 26)

I'd like to move this class and its tests into UT project.

Reply:
Just this class? What about `RecursiveComplexTypesE2ETests`? Is there an example of sharing the test model classes between unit tests and E2E tests that I can follow?

Reply:
No. I don't think we have some models sharing between UT and E2E.

---
In reply to: [193266072](https://github.com/OData/webapi/pull/1458#discussion_r193266072) [](ancestors = 193266072)

Reply:
I ended up sharing the models by putting the source file in a common directory under `test` and linking to it from each project.